Add design doc for simplified Secretless configuration (#716)
cyberark/secretless-broker: design/simpler-configuration.md

  - Both handlers and listeners must be named
  - Rather than using keys as names, we require specifying the `name` of many things
- The current configuration conflates credentials and configuration data
- The http handler configuration is inconsistent over whether the name must match the http handler type or whether (as
  is consistent with other handlers) you can use a random name and specify the `type`.
- In expecting users to understand listeners and handlers to write a configuration, we're exposing an implementation detail
  instead of crafting the configuration in terms that will be clear and sensible to the user.
- In the current experience, we never took into account the user's mental model and designed our experience around that

#### Proposed Experience
```yaml
version: "1"
services:
  postgres-db:
    protocol: pg
    listenOn: 0.0.0.0:5432 # can be a socket as well (same name for both)
    credentials:
      address: postgres.my-service.internal:5432
      password:
        providerId: name-in-vault
        provider: vault
      username:
        providerId: username
        provider: env
    config:  # this section usually blank
      optionalStuff: blah
      

>   # the aws prefix on the credentials indicates which protocol implementation to use

Author: ```suggestion
  # the `aws` type specified in the config indicates which protocol implementation to use
```

  aws-client:
    protocol: http
    listenOn: /var/docker/docker.sock
    credentials:
      accessKeyID:
        providerId: name-in-vault
        provider: conjur
      secretAccessKey:
        providerId: name-in-vault
        provider: conjur
      accessToken:
        providerId: name-in-vault
        provider: conjur
    config:
      type: aws
      pattern: ^http.*

>   # the conjur prefix on the credentials indicates which protocol implementation to use

Author: ```suggestion
  # the `conjur` type specified in the config indicates which protocol implementation to use
```

      pattern: ^http://srdjan.com*

  ssh-handler:
    protocol: ssh
    listenOn: 0.0.0.0:2222
    credentials:
      address: "localhost"
      user: "Jonah"
      privateKey:
        providerId: SSH_PRIVATE_KEY
        provider: env
```
**Note**: In this iteration we are **not** supporting requests for multiple http handler types flowing through a
single service definition. We plan to add support for this in a later iteration. For now, we only support one http handler configured on a given port / socket.

This proposal:
- Integrates the port / socket that Secretless is listening on with how to process the connection (eg credentials, etc)
- It is simple and easy to parse, with everything for a given connection in one small snippet of YAML
- It leverages keys as names to minimize the number of lines required to configure a given handler
- It enables separating configuration and credential data
- It will reduce the options available for configuring the http handler to simplify the experience
- It does not expose the concepts of listeners and handlers, but rather uses naming conventions that will do a better job
  of mapping to the user's mental model of what the broker is doing.

> 

Author: ```suggestion

One other difference is that we enable less verbose options for specifying credentials. For almost all credential providers (Conjur, Kubernetes Secrets, HashiCorp Vault, environment, file, keychain) the syntax is:
```yaml
credentials:
  secretKey:
    providerId: path-to-secret-in-provider
    provider: credentialProvider
```
In the example snippet above, `secretKey` must match a key in the handler configuration (eg `address` for SSH) and `credentialProvider` must match the handle of the desired credential provider (eg `kubernetes` for Kubernetes Secrets). `path-to-secret-in-provider` is the fully qualified ID of the secret in the specified credential store.

The one exception to this syntax is the `literal` provider, which no longer ever needs to be referenced by its `credentialProvider` handle but instead is invoked by simply providing the `secretKey` and the string value that the key should be set to:
```yaml
credentials:
  secretKey: "my-secret-value"
```

```

  instead of crafting the configuration in terms that will be clear and sensible to the user.
- In the current experience, we never took into account the user's mental model and designed our experience around that

#### Proposed Experience
```yaml
version: "1"
services:
  postgres-db:
    protocol: pg
    listenOn: 0.0.0.0:5432 # can be a socket as well (same name for both)
    credentials:
      address: postgres.my-service.internal:5432
      password:
        providerId: name-in-vault
        provider: vault
      username:
        providerId: username
        provider: env
    config:  # this section usually blank
      optionalStuff: blah
      
  # the aws prefix on the credentials indicates which protocol implementation to use
  aws-client:
    protocol: http
    listenOn: /var/docker/docker.sock
    credentials:
      accessKeyID:
        providerId: name-in-vault
        provider: conjur
      secretAccessKey:
        providerId: name-in-vault
        provider: conjur
      accessToken:
        providerId: name-in-vault
        provider: conjur
    config:

>       type: aws

Author: ```suggestion
      authStrategy: aws
```

>       pattern: ^http.*

Reviewer: We had actually changes `pattern` something that gives it the necessary context.  `pattern` just tells you that it's a regex, but a regex for what?  What is its purpose? 
 `requestMatching` or, perhaps even better,`applyToRequestsMatching` tells you what it's doing.

Yes, yes, it's too long!  It's ugly.  But hear me out for a sec...

While I would never agree to name a method `applyToRequestsMatching`, I want to argue that the yml file context is different.  We don't have to type it over and over, and we don't have to use it to reason about object interactions.  Our only real goal here is to make it fully clear to someone reading the file, so that they know what it is _without consulting the docs_.  

Open to improvement, of course, but I do think we should find something better than `pattern`.

  # the conjur prefix on the credentials indicates which protocol implementation to use
  conjur-client:
    protocol: http
    listenOn: 127.0.0.1:8080
    credentials:
      accessToken:
        providerId: /path/to/file
        provider: file
      forceSSL:
        providerId: name-in-vault
        provider: conjur
    config:

>       type: conjur

Author: ```suggestion
      authStrategy: conjur
```

single service definition. We plan to add support for this in a later iteration. For now, we only support one http handler configured on a given port / socket.

This proposal:
- Integrates the port / socket that Secretless is listening on with how to process the connection (eg credentials, etc)
- It is simple and easy to parse, with everything for a given connection in one small snippet of YAML
- It leverages keys as names to minimize the number of lines required to configure a given handler
- It enables separating configuration and credential data
- It will reduce the options available for configuring the http handler to simplify the experience
- It does not expose the concepts of listeners and handlers, but rather uses naming conventions that will do a better job
  of mapping to the user's mental model of what the broker is doing.

One other difference is that we enable less verbose options for specifying credentials. For almost all credential providers (Conjur, Kubernetes Secrets, HashiCorp Vault, environment, file, keychain) the syntax is:
```yaml
credentials:
  secretKey:
    providerId: path-to-secret-in-provider
    provider: credentialProvider
```

> In the example snippet above, `secretKey` must match a key in the handler configuration (eg `address` for SSH) and `credentialProvider` must match the handle of the desired credential provider (eg `kubernetes` for Kubernetes Secrets). `path-to-secret-in-provider` is the fully qualified ID of the secret in the specified credential store.

Author: ```suggestion
In the example snippet above, `secretKey` must match a key in the handler configuration (eg `address` for SSH) and `credentialProvider` must match the unique identifier of the desired credential provider (eg `kubernetes` for Kubernetes Secrets). `path-to-secret-in-provider` is the fully qualified ID of the secret in the specified credential store.
```

> The one exception to this syntax is the `literal` provider, which no longer ever needs to be referenced by its `credentialProvider` handle but instead is invoked by simply providing the `secretKey` and the string value that the key should be set to:

Author: ```suggestion
The one exception to this syntax is the `literal` provider, which no longer ever needs to be referenced by its `credentialProvider` identifier but instead is invoked by simply providing the `secretKey` and the string value that the key should be set to:
```
